### DISCUSSION TITLE
Run a GitHub action on `pull_request` for PR opened from a forked repo

### DISCUSSION THREAD

**Discussion by pvdlg** [community/community#25012] [Actions]
<p>If I create a workflow with an action that runs on <code>pull_request</code>, when a pull request is opened on that repo from a forked repo, the action is not triggered.</p>
<p>Workflow file:</p>
<pre><code class="lang-auto">
workflow "List environment variables" {  
on = "pull\_request"  
resolves = ["my-action"]  
}

action "my-action" {  
uses = "./"  
}

</code></pre>
<p>Then fork the repo, create a branch, push a commit and open a pull request on the upstream repo. The action is not started for this PR.</p>
<p>Is that intentionnal or is it a limitation of the beta?</p>
<p>If that case is planned to be supported in the future, will the secret environment variables be available to the action?</p>
<p>Thanks!</p>

**Comment by appleton** [comment]
<p>Hi <a class="mention" href="/pvdlg">@pvdlg</a>, that’s intentional and likely to remain the case.</p>
<p>This is a mitigation against the possibility that a bad actor could open PRs against your repo and do things like list out secrets or just run up a large bill (once we start charging) on your account. The actions are in fact executed, but it happens against the _fork_, not against the base repo.</p>
<p>This does require that the forked repo also has actions enabled though. During the beta period that means that the owner of the forked repo must also be in the beta.</p>
<p>Hope this helps clear things up!</p>

**Comment by larshp** [comment]
<p>Hi,</p>
<p>Is this documented somewhere?</p>
<p>If I setup <a href="https://github.com/jessfraz/shaking-finger-action" rel="nofollow noopener">https://github.com/jessfraz/shaking-finger-action</a></p>
<p>on <em>pull_request</em>, will the action result show up if somebody with forked repo opens a PR to my repository? Typically I’d like unit tests + linting results to show in the PR automatically.</p>
<p>If it runs against the forked repo, then I assume it is not possible to use the GITHUB_TOKEN to post a comment like in the shaking-finger-action example?</p>

**Comment by AlekSi** [comment]
<p><a href="https://developer.github.com/actions/creating-workflows/workflow-configuration-options/#pull-request-events-for-forked-repositories" rel="nofollow noopener">https://developer.github.com/actions/creating-workflows/workflow-configuration-options/#pull-request-events-for-forked-repositories</a></p>

**Comment by huynguyendev** [comment]
<p>Has there been a change in this policy? I created a test account (<a class="mention" href="/huynguyendev">@huynguyendev</a>) that is not in the beta. I then opened up a PR to TextureGroup/Texture (which is in the beta) from my forked repo. The commit changes Texture’s worflow file to run on pull_request and voila, the workflow is started for my PR!</p>
<p><a href="https://github.com/TextureGroup/Texture/pull/1628/checks?sha=4f0f4cabd6cbb40224329446e100701ca448353d" rel="nofollow noopener">https://github.com/TextureGroup/Texture/pull/1628/checks?sha=4f0f4cabd6cbb40224329446e100701ca448353d</a></p>

**Comment by kirihack13** [comment]
<p>Hi <a class="mention" href="/appleton">@appleton</a> ,</p>
<p>I understand the same. But i think it would be better if atleast the status of the run is shown on the main repository.</p>
<p>The actions run at personal fork but the final status results are shown at the main repository.</p>

**Comment by cclauss** [comment]
<blockquote>
<p>actions are in fact executed, but it happens against the _fork_</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<p>How do I (the owner of the fork) see the results of those actions?</p>
<blockquote>
<p>This does require that the forked repo also has actions enabled</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<p>Is there something manual that needs to be done on the forked repo to enable actions?</p>
<blockquote>
<p>During the beta period that means that the owner of the forked repo must also be in the beta.</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<p>I am in the beta but I can not see the results even when I have <strong>on: [pull_request, push]</strong> set on a pull request in a forked repo. What am I missing?</p>

**Comment by chrispat** [comment, marked as answer]
<p>For the recent update for GitHub Actions we have made a lot of changes to how workflows are run.</p>
<p>A workflow that specifies a pull_request as the event will always run in the repository that contains that pull_request.  So when I raise a pull_request to the upstream repo from a fork of that repo the workflow run will be in that upstream repo.  </p>
<p><a href="https://help.github.com/en/articles/events-that-trigger-workflows#pull-request-events-for-forked-repositories" rel="nofollow noopener">https://help.github.com/en/articles/events-that-trigger-workflows#pull-request-events-for-forked-repositories</a></p>
<p>During the beta you may or may not have GitHub Actions enabled in your user account or organization.  If you don’t have GitHub Actions enabled forking a repo that has actions workflows into your account will not cause any runs to occur.</p>

**Reply by amithm7** [reply]
But a "Fork Pull Request Workflow", how to push back to `fork` from a the workflow?

Even in private repos which can have `GITHUB_TOKEN` with `write` access ([_Send write tokens to workflows from pull requests_](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories)) are only able to write to original repo instead of the fork repo.

**Comment by mpdude** [comment]
<p>So how does the mitigation work in this case?</p>
<p>How can I be sure that the author of the PR did not modify the workflow definition and e. g. reads out repository secrets?</p>

**Comment by mpdude** [comment]
<p>I tried this with two accounts that both have Actions enabled:</p>
<h1>Create repo for user A</h1>
<h1>Fork into B’s account</h1>
<h1>Add a branch, create a PR back into A</h1>
<p>… then a <code>push</code> action will run in B, but the results will not be shown in the PR at A.</p>
<p>Or am I misunderstanding something?</p>

**Comment by slawekjaranowski** [comment]
<p>Secret variables aren’t available on PR by default.</p>
<p>Here is discussion about possibility to pass secret variable to PR: <a href="https://github.community/t/make-secrets-available-to-builds-of-forks/16166">https://github.community/t5/GitHub-Actions/Make-secrets-available-to-builds-of-forks/m-p/30678#M508</a></p>

**Comment by j0k3r** [comment]
<p>Same problem.</p>
<p>Both account with GitHub Action CI enabled but nothing is triggered on the upstream when a PR is made from a fork…</p>

**Comment by ibakshay** [comment]
<p>Hi <a class="mention" href="/larshp">@larshp</a>, are you able to configure the workflow so that unit tests + linting results is shown on the PR when the PR is coming from the forked repo ? … </p>

**Comment by Mathspy** [comment]
<p>Can someone elaborate on whether the marked solution is the current behaviour or the future intended behaviour?</p>
<p>Creating a PR from a fork that has Actions enabled currently does not trigger an actions workflow run on upstream but instead triggers the workflow run on the fork</p>

**Comment by jamienoss** [comment]
<p>Hi,</p>
<p>Is this true for private repos?</p>
<p>No matter which combination of advice I take I cannot get my workflow to run from an upstream PR when the branch being PR’d is from a fork (though the PR is open in the upstream repo). I use a workflow for CI to test the PR’d code.</p>
<p>I see that the <a href="https://help.github.com/en/articles/events-that-trigger-workflows#pull-request-events-for-forked-repositories" rel="nofollow noopener">docs</a> state:</p>
<p><strong>"Note:</strong>  Workflows do not run on private base repositories when you open a pull request from a forked repository."</p>
<p>What am I not understanding as this seems ironic to me given that permissions have to be granted to even gain access to the repo? I.e. if I’m ok with them having access, why should I not be ok with them using a workflow?</p>
<p>Thanks</p>

**Comment by FabianKoestring** [comment]
<p>We use local action runners. Does this mean, that every person who has a fork of that repository has to set his own local action runner?</p>

**Comment by comaniac** [comment]
<p>Same confusion here!</p>
<p>I have manually added “collaborators” to my private repo, so I should trust them to send PRs back!</p>
<p>Now I even cannot get the CI on a forked repo working…</p>
<p>Did anyone figure out a workaround for this annoying issue? The only solution I can think of is 1) asking collaborators to send PRs to a dev branch, 2) I merge PRs to trigger CI with a “push to dev” event, and 3) I rebase master onto dev regularly.</p>

**Comment by nyurik** [comment]
<p>See <a href="https://github.com/nyurik/auto_pr_comments_from_forks" rel="nofollow noopener">https://github.com/nyurik/auto_pr_comments_from_forks</a> – a workaround for this limitation.  Essentially it’s a cron job that frequently checks for recent runs, and posts an artifact content that run has created.</p>

**Comment by tehpsalmist** [comment]
<p>It has been several months and this problem still persists, evidently. Any news on whether this feature will be working for private repos soon?</p>

**Comment by zomglings** [comment]
<p>This is how I am creating comments on PRs from forks this using the new <code>pull_request_target</code> event released by GitHub in August 2020:<br>
</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/simiotics/locust/blob/1bebcc2cac67e90c1efb09b42c37e91cdae829f9/.github/workflows/locust.yaml" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/simiotics/locust/blob/1bebcc2cac67e90c1efb09b42c37e91cdae829f9/.github/workflows/locust.yaml" target="_blank" rel="noopener nofollow ugc">simiotics/locust/blob/1bebcc2cac67e90c1efb09b42c37e91cdae829f9/.github/workflows/locust.yaml</a></h4>
<pre><code class="lang-yaml">name: Locust summary

on: [ pull_request_target ]

jobs:
  build:
    runs-on: ubuntu-20.04
    steps:
      - name: PR head repo
        id: head_repo_name
        run: |
          HEAD_REPO_NAME=$(jq -r '.pull_request.head.repo.full_name' "$GITHUB_EVENT_PATH")
          echo "PR head repo: $HEAD_REPO_NAME"
          echo "::set-output name=repo::$HEAD_REPO_NAME"
      - name: Checkout git repo
        uses: actions/checkout@v2
        with:
          repository: ${{ steps.head_repo_name.outputs.repo }}
          fetch-depth: 0
      - name: Generate Locust summary
</code></pre>

  This file has been truncated. <a href="https://github.com/simiotics/locust/blob/1bebcc2cac67e90c1efb09b42c37e91cdae829f9/.github/workflows/locust.yaml" target="_blank" rel="noopener nofollow ugc">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<hr>
<p>PS:<br>
Dear GitHub team,</p>
<p>The documentation on pull_request_target and workflow_run events is so confusing. The documentation says that both events can be used to create comments on PRs from forks.</p>
<p>I found workflow_run to be completely insufficient to the task, and there were no examples of how to do it for pull_request_target.</p>
<p>Please consider adding examples when you make claims like that on the documentation. It will save a lot of people a decent amount of time.</p>

**Reply by hlindberg** [reply]
Super Thanks! You saved my day, I have been struggling with this and with this info I finally got it to work!
